### PR TITLE
L2 genesis generation w/ predeploy

### PR DESCRIPTION
**Pull request by LEAFERx**
# Goals of PR

Core changes:

- Updated magi to the latest develop branch
- Introduce a new genesis generation tool that can set up predeploys and proxies at the genesis block
  - Adapted from Optimism's `op-bindings` and `op-chain-ops` packages
  - `ops/bindings`: Predeploy and proxy contract bindings generation with deployed bytecode and storage layout
  - `ops/cmd/genesis`: Introduce a new command which reads a `GenesisConfig` (see `ops/genesis/config.go`) and generate a geth genesis file by deploying predeploys on a simulated backend and transform deployment information into genesis file.
- Minor contracts/code clean-up

TODOs:

- [x] Update scripts to use this new tool
- [x] Clean up READMEs, comments and make proper attributions to Optimism
- [ ] Fix forge tests due to use of predeploy addresses in contracts

### CODE REVIEW

File: ops/deployer/deployer.go [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+		ArrowGlacierBlock:   big.NewInt(0),
+		GrayGlacierBlock:    big.NewInt(0),
+		ShanghaiTime:        nil,
+		CancunTime:          nil,
+		// Activated proof of stake. We manually build/commit blocks in the simulator anyway,
+		// and the timestamp verification of PoS is not against the wallclock,
+		// preventing blocks from getting stuck temporarily in the future-blocks queue, decreasing setup time a lot.
+		MergeNetsplitBlock:            big.NewInt(0),
+		TerminalTotalDifficulty:       big.NewInt(0),
+		TerminalTotalDifficultyPassed: true,
```
Comment (MoonShiesty) [inline ops/deployer/deployer.go:73]: should we set `"enableL2GasLimitApi": true` here?

**Review by MoonShiesty — APPROVED** [review]
looks good. optimism creates the deployment artifacts for the L1 and L2 contracts with solidity, is this a follow-up we'd want to implement?

https://github.com/ethereum-optimism/optimism/blob/98ee7d5dbe82c0199b4e4201d49020a6a3ebf1f0/packages/contracts-bedrock/scripts/Deploy.s.sol

File: ops/genesis/genesis.go [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+		IstanbulBlock:                 big.NewInt(0),
+		MuirGlacierBlock:              big.NewInt(0),
+		BerlinBlock:                   big.NewInt(0),
+		LondonBlock:                   big.NewInt(0),
+		ArrowGlacierBlock:             big.NewInt(0),
+		GrayGlacierBlock:              big.NewInt(0),
+		MergeNetsplitBlock:            big.NewInt(0),
+		TerminalTotalDifficulty:       big.NewInt(0),
+		TerminalTotalDifficultyPassed: true,
+		CancunTime:                    nil, // no Dencun on L2 yet.
```
Comment (chalkandpaste) [inline ops/genesis/genesis.go:48]: Same as Julian's comment re: "enableL2GasLimitApi"... is that required here?

**Review by chalkandpaste — APPROVED** [review]
Lightly reviewed the OP code, but everything else LGTM. Let me know if you want me to look in more detail at the OPs code.

**Comment by LEAFERx** [comment]
> looks good. optimism creates the deployment artifacts for the L1 and L2 contracts with solidity, is this a follow-up we'd want to implement?
> 
> https://github.com/ethereum-optimism/optimism/blob/98ee7d5dbe82c0199b4e4201d49020a6a3ebf1f0/packages/contracts-bedrock/scripts/Deploy.s.sol

Yes ideally we want to shift most complex bash scripts to golang/python and consolidate all configurations/deployment in one program

File: ops/.gitignore [hunk 3]
```diff
@@ -0,0 +1,3 @@
+build
+bindings/bindings/*.go
+!bindings/bindings/registry.go
```
Comment (ujvl) [inline ops/.gitignore:3]: nit: whats the `!` and is this file not covered by the ignored path above

**Pull request merged**